ime/app/src/main/java/com/anysoftkeyboard/keyboards/views/PointerTracker.java

```diff
@@ -466,15 +466,41 @@ private void detectAndSendKey(int index, int x, int y, long eventTime) {
                 listener.onCancel();
             }
         } else {
-            if (key.text != null && !mKeyDetector.isKeyShifted(key)) {
+            if (key.text != null || (key.shiftedText != null && mKeyDetector.isKeyShifted(key))) {
                 if (listener != null) {
-                    listener.onText(key, key.text);
-                    listener.onRelease(0); // dummy key code
-                }
-            } else if (key.shiftedText != null && mKeyDetector.isKeyShifted(key)) {
-                if (listener != null) {
-                    listener.onText(key, key.shiftedText);
-                    listener.onRelease(0); // dummy key code
+                    CharSequence text = null;
+                    if (!mKeyDetector.isKeyShifted(key)) {
+                        text = key.text;
+                    } else if (mKeyDetector.isKeyShifted(key)) {
+                        text = key.shiftedText;
+                    }
+                    int[] nearByKeyCodes = mKeyDetector.newCodeArray();
+                    mKeyDetector.getKeyIndexAndNearbyCodes(x, y, nearByKeyCodes);
+                    mTapCount = 0;
+                    for (int i = 0; i < text.length(); i++) {
+                        int code = text.charAt(i);
+                        /*
+                         * Swap the first and second values in the mCodes array if the primary code is not
+                         * the first value but the second value in the array. This happens when key
+                         * debouncing is in effect.
+                         */
+                        if (nearByKeyCodes.length >= 2
+                                && nearByKeyCodes[0] != code
+                                && nearByKeyCodes[1] == code) {
+                            nearByKeyCodes[1] = nearByKeyCodes[0];
+                            nearByKeyCodes[0] = code;
+                        }
+                        if (listener != null) {
+                            if (isInGestureTyping()) {
```

Comment:
I'm not sure I understand this part. Can you add some comments?

Reply:
I'm afraid I'm not 100% on it either, as I just copied and reordered the code a little further down.

Reply:
we need some unit-tests here so we'll know how it behaves.

Reply:
I think this should be called once, before the loop. Again, I'm not really sure what this code does...